src/Campaigns/Actions/RedirectLegacyCreateFormToCreateCampaign.php

```diff
@@ -33,30 +38,73 @@ public function __invoke()
      */
     private function isAddingNewForm(): bool
     {
-        global $pagenow;
+        return $GLOBALS['pagenow'] === 'post-new.php'
+               && isset($_GET['post_type'])
+               && $_GET['post_type'] === 'give_forms';
+    }
 
-        $isOptionBasedFormEditorPage = $pagenow === 'post-new.php';
-        $isVisualFormBuilderPage = $pagenow === 'edit.php' && isset($_GET['page']) && 'givewp-form-builder' === $_GET['page'];
-        $isGiveFormsCpt = isset($_GET['post_type']) && $_GET['post_type'] === 'give_forms';
+    /**
+     * @unreleased
+     */
+    private function isEditingCampaignForm(): bool
+    {
+        return $GLOBALS['pagenow'] === 'edit.php'
+               && isset($_GET['post_type'])
+               && $_GET['post_type'] === 'give_forms'
+               && isset($_GET['page']) && $_GET['page'] === 'givewp-form-builder';
+    }
 
-        return ($isOptionBasedFormEditorPage || $isVisualFormBuilderPage) && $isGiveFormsCpt;
+    /**
+     * @unreleased
+     */
+    private function isP2P(): bool
+    {
+        return isset($_GET['p2p']);
     }
 
     /**
      * @unreleased
      */
-    private function isEditingCampaignForm(): bool
+    private function redirectToNewCampaignPage(): void
+    {
+        wp_safe_redirect(admin_url('edit.php?post_type=give_forms&page=give-campaigns&new=campaign'));
+        exit;
+    }
+
+    /**
+     * @unreleased
+     */
+    private function redirectToP2PNewCampaignPage(): void
     {
-        global $pagenow;
+        wp_safe_redirect(admin_url('edit.php?post_type=give_forms&page=p2p-add-campaign'));
+        exit;
+    }
 
-        $formId = $pagenow === 'post.php' && isset($_GET['post']) ? absint($_GET['post']) : 0;
-        $formId = $pagenow === 'edit.php' && isset($_GET['donationFormID'], $_GET['page']) && 'givewp-form-builder' === $_GET['page'] ? absint($_GET['donationFormID']) : $formId;
-        $isGiveFormsCpt = (isset($_GET['post_type']) && $_GET['post_type'] === 'give_forms') || (get_post_type($formId) === 'give_forms');
+    /**
+     * @unreleased
+     */
+    private function isCampaignIdInvalidOrMissing(): bool
+    {
+        return ! isset($_GET['campaignId']) || ! (Campaign::find(absint($_GET['campaignId'])));
+    }
 
-        if ($formId && $isGiveFormsCpt) {
-            return (bool)Campaign::findByFormId($formId);
+    /**
+     * @unreleased
+     */
+    private function isP2PCampaignFormIdInvalidOrMissing(): bool
+    {
+        if (class_exists(\GiveP2P\P2P\Repositories\CampaignRepository::class)) {
+            return ! isset($_GET['donationFormID']) || ! give(\GiveP2P\P2P\Repositories\CampaignRepository::class)->getByFormId($_GET['donationFormID']);
```

Comment:
@alaca do we need to pull in the p2p repository or can we just check if the campaign type is p2p?

Reply:
@jonwaldstein P2P doesn't use `campaign_type` column.  We added that to the core.

Reply:
What about this migration that adds the p2p type though?  When someone upgrades to this version of Give that should be available to use right?

https://github.com/impress-org/givewp/blob/b13f5ff2706d435b573afb770fcdf2d67cb021f0/src/Campaigns/Migrations/P2P/SetCampaignType.php#L51